utils/treesitter/grammar.js

```diff
@@ -289,51 +294,54 @@ module.exports = grammar({
       ),
 
     where_clause: ($) =>
-      prec(
-        PREC.WhereClause,
-        choice(
-          seq($._expression, '==', $._expression),
-          seq($._expression, 'impls', $._expression),
-          seq($._expression, '=', $._expression),
-          // TODO: Fix conflict with logical and
-          prec.left(seq($.where_clause, 'and', $.where_clause))
+      choice(
+        seq($._simple_expression, '==', $._simple_expression),
+        seq($._simple_expression, 'impls', $._simple_expression),
+        seq($._simple_expression, '=', $._simple_expression),
+        prec.left(
+          PREC.WhereClause + 1,
+          seq($.where_clause, 'and', $.where_clause)
         )
       ),
 
     where_expression: ($) =>
-      prec.left(PREC.TermPostfix, seq($._expression, 'where', $.where_clause)),
+      prec.left(PREC.WhereClause, seq($._expression, 'where', $.where_clause)),
 
     call_expression: ($) =>
-      prec(PREC.TermPostfix, seq($._expression, $.paren_expression)),
+      prec(PREC.TermPostfix, seq($._simple_expression, $.paren_expression)),
 
     pointer_expression: ($) =>
-      prec(PREC.TypePostfix, seq($._expression, $.postfix_star)),
+      prec(PREC.TypePostfix, seq($._simple_expression, $.postfix_star)),
 
-    _expression: ($) =>
+    _simple_expression: ($) =>
       choice(
         $.array_literal,
-        $.as_expression,
-        $.binary_expression,
         $.builtin_type,
         $.call_expression,
         $.deref_expression,
         $.fn_type_expression,
         $.ident,
-        $.if_expression,
         $.index_expression,
         $.literal,
         $.paren_expression,
         $.pointer_expression,
         $.postfix_expression,
         $.ref_expression,
-        $.unary_prefix_expression,
-        $.where_expression,
         'self',
-        // TODO: Remove these two once `where` clauses don't use the expression rule
         '.Self',
         $.designator
```

Comment:
We can't remove this because `designator` can be inside a `paren_expression`

example:
https://github.com/carbon-language/carbon-lang/blob/de614f37e95e55a9118a86546ef7ae48f15f2760/explorer/testdata/assoc_const/rewrite_interface_params.carbon#L25-L27